core/src/main/java/io/parsingdata/metal/token/TokenRef.java

```diff
@@ -60,22 +64,22 @@ public TokenRef(final String name, final String referenceName, final Encoding en
 
     @Override
     protected Optional<Environment> parseImpl(final String scope, final Environment environment, final Encoding encoding) throws IOException {
-        return lookup(environment.order, referenceName).parse(scope, environment, encoding);
+        return lookup(ImmutableList.create(environment.order), referenceName).computeResult().parse(scope, environment, encoding);
     }
 
-    private Token lookup(final ParseItem item, final String referenceName) {
-        if (item.getDefinition().name.equals(referenceName)) {
-            return item.getDefinition();
+    private SafeTrampoline<Token> lookup(final ImmutableList<ParseItem> items, final String referenceName) {
+        if (items.isEmpty()) { return complete(() -> LOOKUP_FAILED); }
+        final ParseItem item = items.head;
+        if (item.getDefinition().name.equals(referenceName)) { return complete(item::getDefinition); }
```

Comment:
One final thing: We need to do a bit of code style cleanup.

Why is this all one line:

```java
if (item.getDefinition().name.equals(referenceName)) { return complete(item::getDefinition); }
```

But this isn't? (From `While.java`, below.)

```java
        if (!environment.isPresent()) {
            return complete(Util::failure);
        }
```

As you might expect, the second style has my preference. Code style used to be a bit more consistent before this PR :relaxed: